kvpress/presses/simlayerkv_press.py

```diff
@@ -0,0 +1,89 @@
+from dataclasses import dataclass
+
+import torch
+from torch import nn
+from transformers import QuantizedCache
+
+from kvpress.presses.base_press import BasePress
+from kvpress.presses.snapkv_press import SnapKVPress
+
+
+@dataclass
+class SimLayerKVPress(BasePress):
+    """
+    SimLayerKV (https://arxiv.org/abs/2410.13846) uses a layer-wise approach to compression:
+        - layers identified as lazy use the Streaming LLM approach (only initial and recent KV pairs are kept)
+        - other layers use the full KV cache
+
+    To identify lazy layers, the last attention weights are used. If the sum of attention weights of the last tokens
+    over the initial and recent tokens is above the lazy_threshold, the layer is considered lazy.
+
+    Official implementation: https://github.com/sail-sg/SimLayerKV. We use n_last=1 to match SKLV-decode
+    """
+
+    lazy_threshold: float
+    n_last: int = 1
+    n_recent: int = 1024
+    n_initial: int = 4
+
+    def is_lazy(
+        self,
+        module: nn.Module,
+        hidden_states: torch.Tensor,
+        keys: torch.Tensor,
+    ) -> bool:
+        """
+        Compute the average attention weights of the last tokens over the initial and recent tokens.
+        A slight difference with the original implementation is that we
+        """
+
+        attn_weights = SnapKVPress.compute_window_attention(module, hidden_states, keys, self.n_last)
+        attn_weights = attn_weights.mean((0, 1, 2))  # mean over bsz, heads and window size
+        score = attn_weights[: self.n_initial].sum() + attn_weights[-self.n_recent :].sum()
+        return score.item() > self.lazy_threshold
+
+    @property
+    def compression_ratio(self):
+        if hasattr(self, "compression_ratios"):
```

Comment:
We can use `__post_init__` to define `self.compression_ratios = None` there. This makes it more explicit.

Reply:
Also, could you implement a setter method that raises an error? This makes it more explicit that one cannot set the compression ratio.

Reply:
done. I think we should always use the same message, I used: 

```python
    @compression_ratio.setter
    def compression_ratio(self, value):
        raise AttributeError("Cannot set the compression ratio")
```